File: 0911-profitable-schemes/0911-profitable-schemes.py
class Solution:
    def profitableSchemes(self, n: int, minProfit: int, group: List[int], profit: List[int]) -> int:
        nCrimes = len(group)
        dp = [[0 for _ in range(minProfit + 1)] for _ in range(n + 1)]

        for i in range(nCrimes, -1, -1):
            for m in range(n, -1, -1):
                for totalProfit in range(minProfit + 1):
                    if i >= nCrimes: 
                        dp[m][totalProfit] = 1 if totalProfit >= minProfit else 0
                    else:
                        if m >= group[i]:
                            dp[m][totalProfit] += dp[m - group[i]][min(totalProfit + profit[i], minProfit)]
                        
                        dp[m][totalProfit] %= 10 ** 9 + 7


        return dp[n][0]

File: 0911-profitable-schemes/0911-profitable-schemes.cpp
class Solution {
public:
    int profitableSchemes(int N, int minProfit, vector<int>& group, vector<int>& profit) {
        vector<vector<vector<long>>> memo(group.size(), vector<vector<long>>(N + 1, vector<long>(minProfit + 1, -1)));
        int mod = 1e9 + 7;

        function<long(int, int, int)> dp = [&] (int i, int n, int p) {
            if (i >= group.size()) {
                if (p == minProfit) {
                    return 1l;
                } else {
                    return 0l;
                }
            }

            if (memo[i][n][p] != -1) return memo[i][n][p];

            long ans = dp(i + 1, n, p);

            if (group[i] <= n) {
                ans = (ans + dp(i + 1, n - group[i], min(profit[i] + p, minProfit))) % mod;
            }

            memo[i][n][p] = ans;

            return ans;
        };

        return dp(0, N, 0);
    }
};